iceoryx_dds_gateway/source/iox2dds_app/main.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <atomic>
+
+#include <iceoryx_posh/runtime/posh_runtime.hpp>
+#include <iceoryx_utils/cxx/helplets.hpp>
+#include <iceoryx_utils/cxx/optional.hpp>
+#include <ioxdds/gateway/iox2dds.hpp>
+
+class ShutdownManager
+{
+  public:
+    static void scheduleShutdown(int num)
+    {
+        char reason;
+        psignal(num, &reason);
+        m_shutdownFlag.store(true, std::memory_order_relaxed);
+    }
+    static bool shouldShutdown()
+    {
+        return m_shutdownFlag.load(std::memory_order_relaxed);
+    }
+
+  private:
+    static std::atomic_bool m_shutdownFlag;
+    ShutdownManager() = default;
+};
+std::atomic_bool ShutdownManager::m_shutdownFlag(false);
+
+int main(int argc, char* argv[])
+{
+    // Set OS signal handlers
+    signal(SIGINT, ShutdownManager::scheduleShutdown);
+    signal(SIGTERM, ShutdownManager::scheduleShutdown);
+
+    // Start application
+    iox::runtime::PoshRuntime::getInstance("/gateway_iox2dds");
+
+    iox::dds::Iceoryx2DDSGateway<> gateway;
+    auto discoveryThread = std::thread([&gateway] { gateway.discoveryLoop(); });
```

Comment:
I would move this thread creation directly into the gateway. Either directly in the constructor or that you start them with a method like ```startupGateway```. The reason is that the user can then handle the gateway much easier and does not have to take care of the resources. Just imagine what happens when ```gateway``` goes out of scope and the user has forgotten to stop the threads or did stop the threads via ```gateway.shutdown()``` but the operating system delayed that and then we have undefined behavior since these threads are working on ```gateway``` which is no longer existing!

Reply:
This makes sense.

Reply:
Fixed.